src/System.Runtime.Extensions/src/System/Net/WebUtility.cs

```diff
@@ -331,30 +332,32 @@ private static byte[] UrlEncode(byte[] bytes, int offset, int count)
 
             // expand not 'safe' characters into %XX, spaces to +s
             byte[] expandedBytes = new byte[count + cUnsafe * 2];
-            int pos = 0;
+            GetEncodedBytes(bytes, offset, count,expandedBytes);
+            return expandedBytes;
+        }
 
-            for (int i = 0; i < count; i++)
+        private static void GetEncodedBytes(byte[] originalBytes, int startIndex, int count, byte[] newBytes)
+        {
+            int pos = 0;
```

Comment:
For the case where originalBytes and newBytes are the same array, could you add some debug asserts to verify that we never get to the point of overwriting input from originalBytes that hasn't yet been read as we write to newBytes?